src/Sentry.pas

```diff
@@ -0,0 +1,169 @@
+unit Sentry;
+
+interface
+
+  uses
+    SysUtils, 
+    MMSystem,
+    Windows,
+    //
+    Typestuff,
+    MutableObject;
+
+  const
+    MAX_BREADCRUMBS = 50;
+
+  type TSentryMetadata = record
+    isDev: boolean;
+    version: Integer;
+    checksum: Cardinal;
+  end;
+
+  type TSentryBreadcrumb = record
+    timestamp: string;
+    msg: string;
+    data: TMutableObject;
+  end;
+
+  type TSentry = class (TOBject)
+    private
+      _output: Textfile;
+      _metadata: TSentryMetadata;
+      _breadcrumbs: array of TSentryBreadcrumb;
+    published
+      constructor Create;
+      procedure addBreadcrumb(crumb: TSentryBreadcrumb);
+      procedure reportError(err: Exception; msg: string);
+  end;
+
+  function makeBreadcrumb(msg: string): TSentryBreadcrumb; overload;
+  function makeBreadcrumb(msg: string; data: TMutableObject): TSentryBreadcrumb; overload;
+
+var
+  sentryModule: TSentry;
+
+implementation
+
+function metadataToJSON(metadata: TSentryMetadata): string; forward;
+function breadcrumbToJSON(crumb: TSentryBreadcrumb): string; forward;
+
+function makeBreadcrumb(msg: string): TSentryBreadcrumb; overload;
+begin
+  result.timestamp := formatdatetime('YYYY-MM-DD hh:mm:ss', now);
+  result.msg := msg;
+  result.data := TMutableObject.Create(true);
+end;
+
+function makeBreadcrumb(msg: string; data: TMutableObject): TSentryBreadcrumb; overload;
+begin
+  result.timestamp := formatdatetime('YYYY-MM-DD hh:mm:ss', now);
+  result.msg := msg;
+  result.data := data;
+end;
+
+constructor TSentry.Create;
+begin
+  with _metadata do
+  begin
+    isDev := {$IFDEF undebug} false {$ELSE} true {$ENDIF};
+    version := PROG_VER;
+    checksum := datachecksum;
+  end;
+end;
+
+procedure TSentry.addBreadcrumb(crumb: TSentryBreadcrumb);
+var
+  i: Integer;
+begin
+  //can append
+  if length(_breadcrumbs) < MAX_BREADCRUMBS then
+  begin
+    setlength(_breadcrumbs, succ(length(_breadcrumbs)));
+    _breadcrumbs[high(_breadcrumbs)] := crumb;
+
+    exit;
+  end;
+
+  //shift
+  for i := low(_breadcrumbs) to pred(high(_breadcrumbs)) do
+    _breadcrumbs[i] := _breadcrumbs[succ(i)];
+
+  setlength(_breadcrumbs, MAX_BREADCRUMBS);
+  _breadcrumbs[high(_breadcrumbs)] := crumb;
+end;
+
+//TODO: actual JSON writer
+procedure TSentry.reportError(err: Exception; msg: string);
+var
+  json, fname: string;
+  i: Integer;
+begin
+  //open output file
```

Comment:
Most, hogy van API POST akár már azt is használhatná de nem akartam ebbe a releasebe még azt is beszuszakolni.